File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys
input = sys.stdin.readline
N = int(input())
A = N*[0]

for i in range(N):
    A[i] = int(input())
answer = ""
stack = []
num = 1
result = True

for i in range(N):
  su = A[i]
  if su>= num:
    while su>=num:
        stack.append(num)
        num +=1
        answer +="+\n"
    stack.pop()
    answer +="-\n"
  else:
        n = stack.pop()
        if su<n:
            print("NO")
            result = False
            break
        else:
            answer +="-\n"
if result:
    print(answer)

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main{
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int n = Integer.parseInt(br.readLine());
        Stack<Integer> stack = new Stack<>();
        int current = 1;

        for(int i=0;i<n;i++){
            int num = Integer.parseInt(br.readLine());
            while(current<= num){
                stack.push(current);
                sb.append("+\n");
                current++;
            }
            if (!stack.isEmpty() && stack.peek() == num) {
                stack.pop();
                sb.append("-\n");
            } else {
                System.out.println("NO");
                return;
            }
        }
        System.out.print(sb);
    }
}